File: src/vm.c
#include "vm.h"
#include "value.h"
#include <stdbool.h>
#include <stdio.h>

virtual_machine vm;

void init_vm(){
	vm.stack_index = 0;
}

void free_vm(){

}

uint8_t read_instruction(uint8_t* ip){
	*ip += 1;
	return vm.chunk->code[*ip - 1];
}

void push(value value){
	vm.stack[vm.stack_index] = value;
	vm.stack_index += 1;
}

value pop(){
	vm.stack_index -= 1;
	return vm.stack[vm.stack_index];
}

vm_error interpret(chunk* chunk){
	vm.chunk = chunk;
	uint8_t ip = 0;

	for(;;) {
		uint8_t instruction;
		switch (instruction = read_instruction(&ip)) {
			case OP_RETURN: {
				print_value(pop());
				return OK;
			}
			case OP_CONSTANT: {
				value constant = vm.chunk->constants.value[read_instruction(&ip)];
				push(constant);
				break;
			}
			case OP_TRUE:{
				push(BOOL_VALUE(true));
				break;
			}
			case OP_FALSE:{
				push(BOOL_VALUE(false));
				break;
			}
			case OP_NOT:{
				value boolean = pop();
				AS_BOOL(boolean) = !AS_BOOL(boolean);
				push(boolean);
				break;
			}
			case OP_NEGATE:{
				value val = pop();
				switch (val.type) {
					case INT_VAL: AS_INT(val) = -AS_INT(val); break;
					case FLOAT_VAL: AS_FLOAT(val) = -AS_FLOAT(val); break;
					default:{
						printf("Invalid type for NEGATE operation\n");
						return RUN_TIME_ERROR;
					}
				}
				push(val);
				break;
			}
			case OP_ADD:{
				value b = pop();
				value a = pop();
				switch (a.type) {
					case INT_VAL: AS_INT(a) += AS_INT(b); break;
					case FLOAT_VAL: AS_FLOAT(a) += AS_FLOAT(b); break;
					default:{
						printf("Invalid type for ADD operation\n");
						return RUN_TIME_ERROR;
					}
				}
				push(a);
				break;
			}
			case OP_SUB:{
				value b = pop();
				value a = pop();
				switch (a.type) {
					case INT_VAL: AS_INT(a) -= AS_INT(b); break;
					case FLOAT_VAL: AS_FLOAT(a) -= AS_FLOAT(b); break;
					default:{
						printf("Invalid type for SUB operation\n");
						return RUN_TIME_ERROR;
					}
				}
				push(a);
				break;
			}
			case OP_MULT:{
				value b = pop();
				value a = pop();
				switch (a.type) {
					case INT_VAL: AS_INT(a) *= AS_INT(b); break;
					case FLOAT_VAL: AS_FLOAT(a) *= AS_FLOAT(b); break;
					default:{
						printf("Invalid type for MULT operation\n");
						return RUN_TIME_ERROR;
					}
				}
				push(a);
				break;
			}
			case OP_DIV:{
				value b = pop();
				value a = pop();
				switch (a.type) {
					case INT_VAL: AS_INT(a) /= AS_INT(b); break;
					case FLOAT_VAL: AS_FLOAT(a) /= AS_FLOAT(b); break;
					default:{
						printf("Invalid type for DIV operation\n");
						return RUN_TIME_ERROR;
					}
				}
				push(a);
				break;
			}
			case OP_GREATER:{
				value b = pop();
				value a = pop();
				value c = (value){.type=BOOL_VAL};
				switch (a.type) {
					case INT_VAL: AS_BOOL(c) = AS_INT(a) > AS_INT(b); break;
					case FLOAT_VAL: AS_BOOL(c) = AS_FLOAT(a) > AS_FLOAT(b); break;
					default:{
						printf("Invalid type for GREATER operation\n");
						return RUN_TIME_ERROR;
					}
				}
				push(c);
				break;
			}
			case OP_LESSER:{
				value b = pop();
				value a = pop();
				value c = (value){.type=BOOL_VAL};
				switch (a.type) {
					case INT_VAL: AS_BOOL(c) = AS_INT(a) < AS_INT(b); break;
					case FLOAT_VAL: AS_BOOL(c) = AS_FLOAT(a) < AS_FLOAT(b); break;
					default:{
						printf("Invalid type for LESSER operation\n");
						return RUN_TIME_ERROR;
					}
				}
				push(c);
				break;
			}
			case OP_GREATER_EQUAL:{
				value b = pop();
				value a = pop();
				value c = (value){.type=BOOL_VAL};
				switch (a.type) {
					case INT_VAL: AS_BOOL(c) = AS_INT(a) >= AS_INT(b); break;
					case FLOAT_VAL: AS_BOOL(c) = AS_FLOAT(a) >= AS_FLOAT(b); break;
					default:{
						printf("Invalid type for GREATER_EQUAL operation\n");
						return RUN_TIME_ERROR;
					}
				}
				push(c);
				break;
			}
			case OP_LESSER_EQUAL:{
				value b = pop();
				value a = pop();
				value c = (value){.type=BOOL_VAL};
				switch (a.type) {
					case INT_VAL: AS_BOOL(c) = AS_INT(a) <= AS_INT(b); break;
					case FLOAT_VAL: AS_BOOL(c) = AS_FLOAT(a) <= AS_FLOAT(b); break;
					default:{
						printf("Invalid type for LESSER_EQUAL operation\n");
						return RUN_TIME_ERROR;
					}
				}
				push(c);
				break;
			}
			case OP_EQUAL_EQUAL:{
				value b = pop();
				value a = pop();
				value c = (value){.type=BOOL_VAL};
				switch (a.type) {
					case INT_VAL: AS_BOOL(c) = AS_INT(a) == AS_INT(b); break;
					case FLOAT_VAL: AS_BOOL(c) = AS_FLOAT(a) == AS_FLOAT(b); break;
					case BOOL_VAL: AS_BOOL(c) = AS_BOOL(a) == AS_BOOL(b); break;
					default:{
						printf("Invalid type for EQUAL_EQUAL operation\n");
						return RUN_TIME_ERROR;
					}
				}
				push(c);
				break;
			}
			case OP_NOT_EQUAL:{
				value b = pop();
				value a = pop();
				value c = (value){.type=BOOL_VAL};
				switch (a.type) {
					case INT_VAL: AS_BOOL(c) = AS_INT(a) != AS_INT(b); break;
					case FLOAT_VAL: AS_BOOL(c) = AS_FLOAT(a) != AS_FLOAT(b); break;
					case BOOL_VAL: AS_BOOL(c) = AS_BOOL(a) != AS_BOOL(b); break;
					default:{
						printf("Invalid type for NOT_EQUAL operation\n");
						return RUN_TIME_ERROR;
					}
				}
				push(c);
				break;
			}
			case OP_AND:{
				value b = pop();
				value a = pop();
				switch (a.type) {
					case BOOL_VAL: AS_BOOL(a) = AS_BOOL(a) && AS_BOOL(b); break;
					default:{
						printf("Invalid type for AND operation\n");
						return RUN_TIME_ERROR;
					}
				}
				push(a);
				break;
			}
			case OP_OR:{
				value b = pop();
				value a = pop();
				switch (a.type) {
					case BOOL_VAL: AS_BOOL(a) = AS_BOOL(a) || AS_BOOL(b); break;
					default:{
						printf("Invalid type for OR operation\n");
						return RUN_TIME_ERROR;
					}
				}
				push(a);
				break;
			}
			default:{
				printf("Unknow instruction in VM\n");
				return RUN_TIME_ERROR;
			}
		}
	}
}

File: src/_vm.zig
const std = @import("std");

const ByteCodePool = @import("bytecode.zig").ByteCodePool;
const ByteCode = @import("bytecode.zig").ByteCode;
const ValueType = @import("bytecode.zig").ValueType;
const Value = @import("bytecode.zig").Value;

pub const VM = struct {
    pool: ByteCodePool,
    stack: std.ArrayList(Value),
    ip: u32 = 0,

    pub fn init(allocator: std.mem.Allocator, pool: ByteCodePool) VM {
        return .{
            .pool = pool,
            .stack = std.ArrayList(Value).init(allocator),
        };
    }

    pub fn deinit(vm: *VM) void {
        vm.stack.deinit();
    }

    fn readInstruction(vm: *VM) ByteCode {
        const bc = vm.pool.bytecodes.items[vm.ip];
        vm.ip += 1;
        return bc;
    }

    pub fn run(vm: *VM) !void {
        while (vm.ip < vm.pool.bytecodes.items.len) {
            const instruction = vm.readInstruction();
            switch (instruction.op_code) {
                .op_return => {
                    // const value = vm.stack.pop();
                    // switch (value) {
                    // .int => |val| std.debug.print("{d}\n", .{val}),
                    // .float => |val| std.debug.print("{d}\n", .{val}),
                    // .boolean => |val| std.debug.print("{any}\n", .{val}),
                    // }
                    return;
                },
                .op_constant => {
                    try vm.stack.append(vm.pool.values.items[instruction.address]);
                },
                .op_not => {
                    var value = vm.stack.pop();
                    switch (value) {
                        .int => std.debug.print("Invalid type for op_not operation\n", .{}),
                        .float => std.debug.print("Invalid type for op_not operation\n", .{}),
                        .boolean => value.boolean = !value.boolean,
                    }
                    try vm.stack.append(value);
                },
                .op_negate => {
                    var value = vm.stack.pop();
                    switch (value) {
                        .int => value.int = -value.int,
                        .float => value.float = -value.float,
                        .boolean => std.debug.print("Invalid type for op_negate operation\n", .{}),
                    }
                    try vm.stack.append(value);
                },
                .op_add => {
                    var b = vm.stack.pop();
                    var a = vm.stack.pop();
                    switch (a) {
                        .int => a.int += b.int,
                        .float => a.float += b.float,
                        .boolean => std.debug.print("Invalid type for op_add operation\n", .{}),
                    }
                    try vm.stack.append(a);
                },
                .op_sub => {
                    var b = vm.stack.pop();
                    var a = vm.stack.pop();
                    switch (a) {
                        .int => a.int -= b.int,
                        .float => a.float -= b.float,
                        .boolean => std.debug.print("Invalid type for op_sub operation\n", .{}),
                    }
                    try vm.stack.append(a);
                },
                .op_mult => {
                    var b = vm.stack.pop();
                    var a = vm.stack.pop();
                    switch (a) {
                        .int => a.int *= b.int,
                        .float => a.float *= b.float,
                        .boolean => std.debug.print("Invalid type for op_mult operation\n", .{}),
                    }
                    try vm.stack.append(a);
                },
                .op_div => {
                    var b = vm.stack.pop();
                    var a = vm.stack.pop();
                    switch (a) {
                        .int => a.int = @divExact(a.int, b.int),
                        .float => a.float /= b.float,
                        .boolean => std.debug.print("Invalid type for op_div operation\n", .{}),
                    }
                    try vm.stack.append(a);
                },
                .op_greater => {
                    var b = vm.stack.pop();
                    var a = vm.stack.pop();
                    var c: Value = .{ .boolean = false };
                    switch (a) {
                        .int => c.boolean = a.int > b.int,
                        .float => c.boolean = a.float > b.float,
                        .boolean => std.debug.print("Invalid type for op_greater operation\n", .{}),
                    }
                    try vm.stack.append(c);
                },
                .op_less => {
                    var b = vm.stack.pop();
                    var a = vm.stack.pop();
                    var c: Value = .{ .boolean = false };
                    switch (a) {
                        .int => c.boolean = a.int < b.int,
                        .float => c.boolean = a.float < b.float,
                        .boolean => std.debug.print("Invalid type for op_less operation\n", .{}),
                    }
                    try vm.stack.append(c);
                },
                .op_greater_than => {
                    var b = vm.stack.pop();
                    var a = vm.stack.pop();
                    var c: Value = .{ .boolean = false };
                    switch (a) {
                        .int => c.boolean = a.int >= b.int,
                        .float => c.boolean = a.float >= b.float,
                        .boolean => std.debug.print("Invalid type for op_greater_than operation\n", .{}),
                    }
                    try vm.stack.append(c);
                },
                .op_less_than => {
                    var b = vm.stack.pop();
                    var a = vm.stack.pop();
                    var c: Value = .{ .boolean = false };
                    switch (a) {
                        .int => c.boolean = a.int <= b.int,
                        .float => c.boolean = a.float <= b.float,
                        .boolean => std.debug.print("Invalid type for op_less_than operation\n", .{}),
                    }
                    try vm.stack.append(c);
                },
                .op_equal => {
                    var b = vm.stack.pop();
                    var a = vm.stack.pop();
                    var c: Value = .{ .boolean = false };
                    switch (a) {
                        .int => c.boolean = a.int == b.int,
                        .float => c.boolean = a.float == b.float,
                        .boolean => c.boolean = a.boolean == b.boolean,
                    }
                    try vm.stack.append(c);
                },
                .op_not_equal => {
                    var b = vm.stack.pop();
                    var a = vm.stack.pop();
                    var c: Value = .{ .boolean = false };
                    switch (a) {
                        .int => c.boolean = a.int != b.int,
                        .float => c.boolean = a.float != b.float,
                        .boolean => c.boolean = a.boolean != b.boolean,
                    }
                    try vm.stack.append(c);
                },
                .op_and => {
                    var b = vm.stack.pop();
                    var a = vm.stack.pop();
                    switch (a) {
                        .int => std.debug.print("Invalid type for op_and operation\n", .{}),
                        .float => std.debug.print("Invalid type for op_and operation\n", .{}),
                        .boolean => a.boolean = a.boolean and b.boolean,
                    }
                    try vm.stack.append(a);
                },
                .op_or => {
                    var b = vm.stack.pop();
                    var a = vm.stack.pop();
                    switch (a) {
                        .int => std.debug.print("Invalid type for op_or operation\n", .{}),
                        .float => std.debug.print("Invalid type for op_or operation\n", .{}),
                        .boolean => a.boolean = a.boolean or b.boolean,
                    }
                    try vm.stack.append(a);
                },
                .op_load_gv => {
                    var a = vm.stack.pop();
                    var key = vm.pool.global_var_tables.values.keys()[instruction.address];
                    vm.pool.global_var_tables.values.put(key, a) catch |err| {
                        std.debug.print("Unable to assign value to variable: {}", .{err});
                    };
                },
                .op_unload_gv => {
                    var value = vm.pool.global_var_tables.values.values()[instruction.address];
                    try vm.stack.append(value);
                },
                .op_print => {
                    const value = vm.stack.pop();
                    switch (value) {
                        .int => |val| std.debug.print("{d}\n", .{val}),
                        .float => |val| std.debug.print("{d}\n", .{val}),
                        .boolean => |val| std.debug.print("{any}\n", .{val}),
                    }
                },
            }
        }
    }
};
